sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -848,11 +967,84 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
         filter
       }
 
+    // Similar to the above Filter over Union
+    // LeftSemi/LeftAnti over Union
+    case join @ Join(union: Union, rightOp, LeftSemiOrAnti(joinType), joinCond) =>
+      if (joinCond.isEmpty) {
+        // Push down the Join below Union
+        val newGrandChildren = union.children.map { grandchild =>
+          Join(grandchild, rightOp, joinType, joinCond)
+        }
+        union.withNewChildren(newGrandChildren)
+      } else {
+        // Union could change the rows, so non-deterministic predicate can't be pushed down
+        val (candidates, containingNonDeterministic) =
+          splitConjunctivePredicates(joinCond.get).span(_.deterministic)
+
+        val (pushDown, rest) = candidates.partition { cond =>
+          !SubqueryExpression.hasCorrelatedSubquery(cond) &&
+            !SubExprUtils.containsOuter(cond)
+        }
+        val stayUp = rest ++ containingNonDeterministic
+
+        if (pushDown.nonEmpty) {
+          val pushDownCond = pushDown.reduceLeft(And)
+          val output = union.output
+          val newGrandChildren = union.children.map { grandchild =>
+            val newCond = pushDownCond transform {
+              case e if output.exists(_.semanticEquals(e)) =>
+                grandchild.output(output.indexWhere(_.semanticEquals(e)))
+            }
+            assert(newCond.references.subsetOf(grandchild.outputSet ++ rightOp.outputSet))
+            Join(grandchild, rightOp, joinType, Option(newCond))
+          }
+          val newUnion = union.withNewChildren(newGrandChildren)
+          if (stayUp.isEmpty) newUnion else Filter(stayUp.reduceLeft(And), newUnion)
+        } else {
+          // Nothing to push down
+          join
+        }
+      }
+
```

Comment:
Similar to the LeftSemi/LeftAnti over Aggregate, pushing down LeftSemi/LeftAnti under a Union means we need to repeat the processing of LeftSemi/LeftAnti once for each branch of the Union.

This code is added to preserve the existing behaviour.